rpki-validator/src/main/java/net/ripe/rpki/validator3/background/BackgroundJobs.java

```diff
@@ -51,72 +46,71 @@
-        schedule(RpkiObjectCleanupJob.class,
-                futureDate(3, MINUTE),
-                simpleSchedule().repeatForever().withIntervalInMinutes(10));
-
-        schedule(RpkiRepositoryCleanupJob.class,
-                futureDate(4, MINUTE),
-                simpleSchedule().repeatForever().withIntervalInMinutes(60));
-
-        schedule(ValidationRunCleanupJob.class,
-                futureDate(5, MINUTE),
-                simpleSchedule().repeatForever().withIntervalInMinutes(5));
-
-        schedule(ValidateRsyncRepositoriesJob.class,
-                futureDate(10, SECOND),
-                simpleSchedule().repeatForever().withIntervalInMinutes(1));
-
-        schedule(DownloadBgpRisDumpsJob.class,
-                futureDate(10, SECOND),
-                simpleSchedule().repeatForever().withIntervalInMinutes(10));
+    public void initBackgroundTasks() {
+        schedule("rpkiObjectCleanupService", 3, 10, TimeUnit.MINUTES,
+            () -> rpkiObjectCleanupService.cleanupRpkiObjects());
+
+        schedule("rpkiRepositoryCleanupService", 4, 60, TimeUnit.MINUTES,
+            () -> rpkiRepositoryCleanupService.cleanupRpkiRepositories());
+
+        schedule("validationRunCleanupService", 5, 5, TimeUnit.MINUTES,
+            () -> validationRunCleanupService.cleanupValidationRuns());
+
+        schedule("rpkiRepositoryValidationService", 10, 60, TimeUnit.SECONDS,
+            () -> rpkiRepositoryValidationService.validateRsyncRepositories());
+
+        schedule("bgpPreviewService", 10, 600, TimeUnit.SECONDS,
+            () -> bgpPreviewService.downloadRisPreview());
     }
 
-    private <T extends Trigger> void schedule(Class<? extends Job> jobClass, Date startAt, ScheduleBuilder<T> schedule) throws SchedulerException {
-        final JobKey jobKey = JobKey.jobKey(CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_HYPHEN, jobClass.getSimpleName()));
-        if (scheduler.checkExists(jobKey)) {
-            scheduler.deleteJob(jobKey);
-        }
-        scheduler.scheduleJob(
-                newJob(jobClass)
-                        .withIdentity(jobKey)
-                        .build(),
-                newTrigger()
-                        .withIdentity(jobKey + "-Trigger")
-                        .startAt(startAt)
-                        .withSchedule(schedule)
-                        .build()
-        );
-        log.info(String.format("Scheduled '%s', starting from '%s'", jobKey, startAt));
+    private void schedule(String key, long delay, long period, TimeUnit timeUnit, Runnable r) {
+        final Runnable rr = () -> {
+            jobToBeExecuted(key);
+            try {
+                r.run();
+            } catch (Exception e) {
+                log.error(String.format("Error executing job '%s'", key), e);
+            } finally {
+                jobWasExecuted(key);
+            }
+        };
+        scheduledExecutorService.scheduleAtFixedRate(rr, delay, period, timeUnit);
```

Comment:
Maybe `scheduleWithFixedDelay` to avoid running jobs concurrently when they take a long time to run?

Reply:
Documentation says it will never happen.